src/main/java/us/hebi/matlab/mat/types/AbstractStruct.java

```diff
@@ -111,4 +111,37 @@ public void close() throws IOException {
     private final List<String> fields = new ArrayList<String>();
     private final List<Array[]> values = new ArrayList<Array[]>();
 
+    @Override
+    protected int subHashCode() {
+        final int prime = 31;
+        int result = 1;
+        result = prime * result + indexMap.hashCode();
+        result = prime * result + fields.hashCode();
+        for (Array[] valueArray : values) {
+            result = prime * result + Arrays.hashCode(valueArray);
+        }
+        return result;
+    }
+
+    protected abstract int subSubHashCode();
+
+    @Override
+    protected boolean subEqualsGuaranteedSameClass(Object otherGuaranteedSameClass) {
+        AbstractStruct other = (AbstractStruct) otherGuaranteedSameClass;
+        if (subSubEquals(otherGuaranteedSameClass) &&
+                other.indexMap.equals(indexMap) &&
+                other.fields.equals(fields) &&
+                other.values.size() == values.size()) {
+            for (int i = 0; i < values.size(); ++i) {
+                if (!Arrays.equals(other.values.get(i), values.get(i))) {
+                    return false;
+                }
+            }
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    protected abstract boolean subSubEquals(Object otherGuaranteedSameClass);
```

Comment:
I can see a reason for the first `sub-` level, but nesting them seems a bit much. I'd prefer children to just override the `subEqualsGuaranteedSameClass` and call `super` if necessary.

If you feel strongly about it and really want to enforce explicit implementation, the `AbstractStruct::subEqualsGuaranteedSameClass` should be `final`.

Reply:
Roger.  Subclassing and equality are tricky, so I like the `final` approach.

Reply:
Nvm, I think you are right that calling `super.` is better, I misunderstood the class hierarchy.